### DISCUSSION TITLE
Support for strongly-typed IDs

### DISCUSSION THREAD

**Discussion by adrianfrankowski95** [JasperFx/marten#2046] [Q&A] [answered]
Hello!

First of all I would like to thank all contributors, I really find this project very unique and interesting!

I encountered an issue regarding strongly-typed IDs. In my project I use DDD and create custom classes for IDs, like in this example:
```
public class PostId : ValueObject<PostId>
{
    public Guid Value { get; }

    public PostId(Guid value)
    {
        Value = value;
    }

    protected override IEnumerable<object> GetEqualityCheckAttributes()
    {
        yield return Value;
    }
}
```

And consume it in a following way:
```
public abstract class Entity<T>
{
    public T Id { get; protected set; }
}

public class PostBase : Entity<PostId>
{
    ...
}
```

Then I'm trying to tell Marten which value is an ID:
```
builder.Services.AddMarten(opt =>
{
    opt.Schema.For<PostBase>()
        .Identity(x => x.Id.Value);
});
```

But unfortunately such setup results in an ArgumentOutOfRangeException:

> System.Private.CoreLib.dll: 'Id members must be an int, long, Guid, or string'

In EF Core I could easily provide convertion to/from GUID:
```
public virtual void Configure(EntityTypeBuilder<PostBase> builder)
{
        builder
            .Property(x => x.Id)
            .HasConversion(x => x.Value, x => new PostId(x))
            .HasMaxLength(68);
}
```

Is there any workaround for this issue or shall I just use pure GUID?

Thank you!

**Comment by jeremydmiller** [comment, marked as answer]
Heh, I just closed the last issue someone opened for strongly typed ids. I'm against doing this in Marten tbh, because I just don't want the extra complexity. You *could* possibly use this as is with Marten by making your strong typed id types be implicitly convertible to the Marten supported identifier types like an int, long, string, or Guid.

**Reply by adrianfrankowski95** [reply]
Thank you for your answer. I've just tried to implement implicit/explicit conversions to/from Guid and unfortunately I still cannot make it work:
```
public static implicit operator Guid(PostId id) => id.Value;
public static implicit operator PostId(Guid id) => new(id);
```
And in registering Marten I tried the following ways:
```
.Identity(x => x.Id);
.Identity(x => x.Id.Value);
.Identity(x => (Guid)x.Id);
```

And I still get the same exception. I also tried to convert it to string with no luck.
Could you please share working example of what kind of conversion is expected?

BTW: I'm using Marten 5.0.0-alpha.4

**Reply by jeremydmiller** [reply]
In that case, I don't think there is a workable solution with the existing Marten and you'll have to revert to using primitive types for the Id.

And in your `Identity(expression)` usage, Marten is only picking up the property or field and doesn't know how to do any kind of magic casting at runtime.
